// ng2-destroy-subscribers/lib/decorator.js
"use strict";
var rxjs_1 = require("rxjs");
function DestroySubscribers(params) {
    return function (target) {
        params = Object.assign({
            addSubscribersFunc: 'addSubscribers',
            removeSubscribersFunc: 'removeSubscribers',
            initFunc: 'ngOnInit',
            destroyFunc: 'ngOnDestroy',
        }, params);
        target.prototype[params.initFunc] = ngOnInitDecorator(target.prototype[params.initFunc]);
        target.prototype[params.destroyFunc] = ngOnDestroyDecorator(target.prototype[params.destroyFunc]);
        function ngOnDestroyDecorator(f) {
            return function () {
                var superData = f ? f.apply(this, arguments) : null;
                if (typeof this[params.removeSubscribersFunc] === 'function') {
                    this[params.removeSubscribersFunc]();
                }
                for (var subscriberKey in this.subscribers) {
                    var subscriber = this.subscribers[subscriberKey];
                    if (subscriber instanceof rxjs_1.Subscriber) {
                        subscriber.unsubscribe();
                    }
                }
                return superData;
            };
        }
        function ngOnInitDecorator(f) {
            return function () {
                var superData = f ? f.apply(this, arguments) : null;
                if (typeof this[params.addSubscribersFunc] === 'function') {
                    this[params.addSubscribersFunc]();
                }
                return superData;
            };
        }
        return target;
    };
}
exports.DestroySubscribers = DestroySubscribers;
